dev/src/bulk-writer.ts

```diff
@@ -648,16 +850,33 @@ export class BulkWriter {
    *   console.log('Executed all writes');
    * });
    */
-  close(): Promise<void> {
+  async close(): Promise<void> {
     this.verifyNotClosed();
     this.firestore._decrementBulkWritersCount();
     const flushPromise = this.flush();
-    this.closed = true;
-    return flushPromise;
+    this.closeCalled = true;
+    await flushPromise;
+
+    // Wait for pending retry operations to run.
+    if (this.batchQueue.length > 0) {
+      await this._flush();
+      await this.closedDeferred.promise;
+    }
```

Comment:
I now see that `verifyNotClosed` only throws if fully closed when called from `retry`. In which case, I think my second question still stands. Do we know that the `await this._flush()` won't cause more retries? Do we either need to do this in a loop or at least await another flush after `this.closeDeferred.promise` resolves?

Reply:
I think we can use Promise chaining here as proposed in the diff above. Instead of requiring `void` as the return type of `onWriteError`, we can require the user to call `retry` and to pass the result of `retry` to use via the error handler. The signature would then be:

`onWriteError((err:BulkWriterError) => Promise<WriteResult>)`

Once we do that, we can chain the result of this error callback and block on the retry attempt in the flush call.

If we don't do this, we have to somehow ensure that all error callbacks have fired once we process the results. This would be a bit tricky and may not port to Java or other languages:

```
error.reject(...);
setImmediate(() => {
  // This only works in Node, but due to the way that ticks are processed in the VM, we can use a `setImmediate`
  // to ensure that our code only runs once all Promise chains are processed.
});
```

Reply:
That would be elegant, I think. The ultimate result of the original write would become the result of the retry. It would just be important to document well so that callers don't forget to return the result of `retry()` from their error handler or the retry promise would end up unhandled and the original write would resolve with undefined.